### PR TITLE
Add support for attached documents

### PR DESCRIPTION
**Pull request by coltfred**
This adds support for decrypting v4 & v5 attached documents. It also adds support for producing v5 attached documents.

I've done a bit of self review, but I'm currently feeling pretty sick so I haven't looked at it intensely yet.

### CODE REVIEW

File: src/standalone/standard.rs [hunk 1]
```diff
@@ -80,16 +83,43 @@ impl StandaloneStandardClient {
...
+            })
+        };
+
+        secrets
+            .into_iter()
+            .map(attempt_decrypt)
+            .find_or_first(|result| result.is_ok())
+            .unwrap_or(Err(AlloyError::InvalidConfiguration(
+                "No secret could be found to decrypt".to_string(),
+            )))
```
Comment (giarc3) [inline src/standalone/standard.rs:122]: Cool find. I don't think the `unwrap_or` is possible, but I guess it's better to catch than expect?

File: src/standalone/standard_attached.rs [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+    async fn get_searchable_edek_prefix(&self, id: i32) -> Vec<u8> {
+        self.standard_client.get_searchable_edek_prefix(id)
+    }
+}
+
+/// Encrypt all the fields of the document, attaching the encryption header to the resulting bytes.
+/// Each field will generate a DEK using a derived key and use it along with a random IV to
+/// encrypt the field.
+pub fn encrypt_field_attached<U: AsRef<[u8]>, R: RngCore + CryptoRng>(
```
Comment (giarc3) [inline src/standalone/standard_attached.rs:59]: This function is unused

File: src/standalone/standard_attached.rs [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    let (aes_dek, v4_doc) = generate_aes_edek_and_sign(rng, per_tenant_kek, None, "")?;
+    let attached_payload = encrypt_document_and_attach_iv(
+        rng,
+        aes_dek,
+        ironcore_documents::aes::PlaintextDocument(field.as_ref().to_vec()),
+    )?;
+    Ok(encode_attached_edoc(v4_doc, attached_payload)?.to_vec())
+}
+
+pub fn decrypt_field_attached(
```
Comment (giarc3) [inline src/standalone/standard_attached.rs:76]: This one too

File: src/standard.rs [hunk 4]
```diff
@@ -201,12 +202,14 @@ pub(crate) fn decrypt_document_core(
...
-            if ciphertext.starts_with(&[V3]) {
+            if ciphertext.starts_with(&v3::VERSION_AND_MAGIC) {
                 let encrypted_payload: v3::EncryptedPayload = ciphertext.try_into()?;
                 encrypted_payload.decrypt(&dek)
-            } else {
+            } else if ciphertext.starts_with(&v5::VERSION_AND_MAGIC) {
                 let encrypted_payload: v5::EncryptedPayload = ciphertext.try_into()?;
                 encrypted_payload.decrypt(&dek)
+            } else {
+                decrypt_document_with_attached_iv(&dek, &IvAndCiphertext(ciphertext.into()))
```
Comment (giarc3) [inline src/standard.rs:212]: Add comment that this is V4

File: src/standard_attached.rs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    ) -> Result<PlaintextBytes, AlloyError>;
+    async fn get_searchable_edek_prefix(&self, id: i32) -> Vec<u8>;
+}
+
+pub(crate) async fn encrypt_core<T: StandardDocumentOps>(
+    standard_client: &T,
+    plaintext_field: Vec<u8>,
+    metadata: &AlloyMetadata,
+) -> Result<EncryptedAttachedDocument, AlloyError> {
+    let hardcoded_key = "".to_string();
```
Comment (giarc3) [inline src/standard_attached.rs:42]: I don't understand this

File: src/util.rs [hunk 6]
```diff
@@ -125,6 +126,12 @@ where
     }
 }
 
+pub(crate) fn v4_proto_from_bytes<B: AsRef<[u8]>>(
+    b: B,
+) -> Result<ironcore_documents::icl_header_v4::V4DocumentHeader, AlloyError> {
+    Ok(Message::parse_from_bytes(b.as_ref())?)
```
Comment (giarc3) [inline src/util.rs:132]: How come this was added?

**Review by giarc3 — CHANGES_REQUESTED** [review]

File: src/standard_attached.rs (hunk 5 above)
Comment (coltfred) [reply src/standard_attached.rs:42]: I've added a comment.

File: src/standalone/standard.rs (hunk 1 above)
Comment (coltfred) [reply src/standalone/standard.rs:122]: Ya, it shouldn't ever be able to happen, but if something went wrong I'd rather see an error than expect.

File: src/saas_shield/standard_attached.rs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        &self,
+        attached_field: crate::standard_attached::EncryptedAttachedDocument,
+        metadata: &crate::AlloyMetadata,
+    ) -> Result<crate::PlaintextBytes, crate::errors::AlloyError> {
+        decrypt_core(&self.standard_client, attached_field, metadata).await
+    }
+
+    async fn get_searchable_edek_prefix(&self, id: i32) -> Vec<u8> {
+        self.standard_client.get_searchable_edek_prefix(id)
+    }
```
[the remark this answers is not in the document]
Comment (coltfred) [reply src/saas_shield/standard_attached.rs:33]: I'll just comment it.

File: src/saas_shield/standard.rs [hunk 8]
```diff
@@ -147,15 +150,20 @@ impl SaasShieldStandardClient {
...
             ))
                 }
             }
-            Err(_) => Ok(EdekParts::V3(encrypted_bytes.0.into())),
+            Err(_) => {
+                let bytes: Bytes = encrypted_bytes.0.into();
+                v4::attached::decode_attached_edoc(bytes.clone())
+                    .map(|(edek, _)| EdekParts::V4(edek))
+                    .or_else(|_| Ok(EdekParts::V3(bytes)))
+            }
```
[the remark this answers is not in the document]
Comment (coltfred) [reply src/saas_shield/standard.rs:166]: Added some comments.

**Review by giarc3 — APPROVED** [review]

**Pull request merged**